cpu tuning and Iothreads allocation GSoC

gqlo opened the pull request:
### What does this PR do?
Making `IOThreads` allocation available.  These are dedicated event loop threads for supported disk devices to perform block I/O requests.  `Iothreads` parameter specify the number of IOThreads to be assigned to the domain by supported target storage devices.

`IOThreads` can be pined using `iothreadpin` parameter  in `cputune`  which provides details regarding the CPU tunable parameters for the domain.  `cputune` contains the following parameters: 

-  `vcpupin`
- `emulatorpin`
-  `iothreadpin`
-   `shares`
- `period`
- `quota`
- `global_period`
- `global_quota`
- `emulator_period`
- `emulator_quota`
- `iothread_period`
- `iothread_quota`
- `vcpusched, iothreadsched and emulatorsched`
- `cachetune`
- `memorytune`

Detailed Libvirt domain xml info can be found at [iothreads](https://libvirt.org/formatdomain.html#iothreads-allocation)  and  [cpu tuning](https://libvirt.org/formatdomain.html#cpu-tuning)

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltstack.com/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [x] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes/No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

gqlo committed:
virt: support cpu tunning and Iothread allocation

cbosdo commented on salt/modules/virt.py:
```diff
@@ -1747,6 +1779,7 @@ def init(
     boot_dev=None,
     numatune=None,
     iothreads=None,
+    cputune=None,
```
since this is adding tuning of the virtual cpu, I would move it in the `cpu` param, maybe a `tuning` key of it.

cbosdo commented on salt/modules/virt.py:
```diff
@@ -3237,7 +3409,6 @@ def _append_placeholder(value_dict, key):
                 if isinstance(item["value"], dict):
                     for k, v in item["value"].items():
                         values.append({"value": v, "sid": k, "id": item["id"]})
-
```
don't remove blank lines

cbosdo commented on salt/modules/virt.py:
```diff
@@ -3258,6 +3429,7 @@ def _append_placeholder(value_dict, key):
                     else ""
                     for placeholder in placeholders
                 }
+
```
likely adding a blank line that shouldn't have been removed in the first place

cbosdo commented on salt/modules/virt.py:
```diff
@@ -2090,6 +2152,70 @@ def init(
+        including vCPU or iothreads will be pinned to.
+
+    iothreadpin
+        The optional iothreadpin element specifies which of host physical CPUs the IOThreads will be pinned to.
```
Mention here that the IOThread IDs to use in this value are from 1 to the number passed to `iothread` parameter.

cbosdo reviewed (changes requested):
See my comments inline. Also please merge the cputune-related commits into one

gqlo committed:
virt: support host numa tunning capability

gqlo committed:
virt: support cpu model and topology

gqlo committed:
virt: support memory_backing

cbosdo commented on salt/modules/virt.py:
```diff
@@ -3213,6 +3377,107 @@ def _set_vcpu(node, value):
+            "set": lambda n, v: n.set("priority", str(v)),
+            "del": salt.utils.xmlutil.del_attribute("priority", ["vcpus"]),
+        },
+        {
+            "path": "cpu:tuning:iothreadsched:iothreads",
+            "xpath": "cputune/iothreadsched",
+            "get": lambda n: str(n.get("iothreads")) if n.get("iothreads") else None,
+            "set": lambda n, v: n.set("iothreads", str(v)),
+            "del": salt.utils.xmlutil.del_attribute("iothreads", []),
+        },
+        {
+            "path": "cpu:tuning:iothreadsched:scheduler",
+            "xpath": "cputune/iothreadsched",
+            "get": lambda n: str(n.get("iothreads")) if n.get("scheduler") else None,
+            "set": lambda n, v: n.set("scheduler", str(v)),
+            "del": salt.utils.xmlutil.del_attribute("scheduler", []),
+        },
```
As per the [schema](https://gitlab.com/libvirt/libvirt/-/blob/master/docs/schemas/domaincommon.rng#L966) there can be mulitple `iothreadsched`... they need to be a list

You can also add the `emulatorsched` (not a list)

cbosdo commented on salt/modules/virt.py:
```diff
@@ -3213,6 +3377,107 @@ def _set_vcpu(node, value):
+            "set": lambda n, v: n.set("scheduler", str(v)),
+            "del": salt.utils.xmlutil.del_attribute("scheduler", []),
+        },
+        {
+            "path": "cpu:tuning:cachetune:{id}",
+            "xpath": "cputune/cachetune[@vcpus='$id']",
+            "get": lambda n: str(n.get("vcpus")) if n.get("vcpus") else None,
+            "set": lambda n, v: n.set("vcpus", str(v)),
+            "del": salt.utils.xmlutil.del_attribute("vcpus", ["id"]),
+        },
```
I think this entry is a bogus one to be removed

cbosdo commented on salt/modules/virt.py:
```diff
@@ -3213,6 +3377,107 @@ def _set_vcpu(node, value):
+            "set": lambda n, v: n.set("type", str(v)),
+            "del": salt.utils.xmlutil.del_attribute("type", ["id"]),
+        },
+        {
+            "path": "cpu:tuning:cachetune:{id}:{sid}:size",
+            "xpath": "cputune/cachetune[@vcpus='$id']/cache[@id='$sid']",
+            "get": lambda n: str(n.get("size")) if n.get("size") else None,
+            "set": lambda n, v: n.set("size", str(v)),
+            "del": salt.utils.xmlutil.del_attribute("size", ["id"]),
+        },
```
Should handle the unit here.

cbosdo commented on salt/states/virt.py:
```diff
@@ -636,6 +750,18 @@ def running(
 
         .. versionadded:: Magnesium
 
+    :param numatune:
+        The optional numatune element provides details of how to tune the performance of a NUMA host via controlling NUMA
+        policy for domain process. The optional ``memory`` element specifies how to allocate memory for the domain process
+        on a NUMA host. ``memnode`` elements can specify memory allocation policies per each guest NUMA node. The definition
+        used in the dictionary can be found at :ref:`init-cpu-def`.
+
+        To update any numatune parameters, specify the new value. To remove any ``numatune`` parameters, pass a None object,
+        for instance: 'numatune': ``None``. Please note that ``None`` is mapped to ``null`` in sls file, pass ``null`` in
+        sls file instead.
+
+        .. versionadded:: Magnesium
+
```
Wrong PR :)

cbosdo reviewed (changes requested):
Left a few comments to address.

Commit:
fixup! precommit failure fix

cbosdo committed:
virt: make context preprocessing more reusable in _gen_xml

Introduce mapping structures in order to help reusing the common patterns
in the virt._gen_xml() context pre processing.

cbosdo commented on salt/modules/virt.py:
```diff
@@ -1959,6 +2003,61 @@ def init(
+                        scheduler: rr
+                        priority: 3
+                    iothreadsched:
+                        iothreads: 2
+                        scheduler: batch
```
This doesn't match the implementation. Should be like this:

```diff
-                        iothreads: 2
-                        scheduler: batch
+                       0:
+                          iothreads: 2
+                          scheduler: batch
```

But even this way, how do we handle the default then where the `iothread id` is not defined? Same question for `vcpusched`

cbosdo commented on salt/modules/virt.py:
```diff
@@ -951,6 +952,48 @@ def _gen_xml(
+                                        for k, v in param.items():
+                                            if v:
+                                                context["cputune"][tag][k] = v
+                                    elif tag in ["vcpupin", "iothreadpin"]:
+                                        for k, v in param.items():
+                                            if v:
+                                                context["cputune"][tag][
+                                                    k
+                                                ] = "{}='{}'".format("cpuset", v)
```
Since the `cpuset` attribute is mandatory in the XML schema, it would make the whole code simpler to move the attribute name to the jinja template and get rid of this piece of code

cbosdo commented on salt/modules/virt.py:
```diff
@@ -951,6 +952,48 @@ def _gen_xml(
+                                    elif tag == "iothreadsched":
+                                        for k, v in param.items():
+                                            if v:
+                                                context["cputune"][tag][
+                                                    k
+                                                ] = "{}='{}'".format(k, v)
```
Same here, using the loop on the attributes as you did for `vcpusched` in the jinja template allows getting rid of this code

cbosdo commented on salt/modules/virt.py:
```diff
@@ -951,6 +952,48 @@ def _gen_xml(
+                                        "iothread_period",
+                                        "iothread_quota",
+                                    ]:
+                                        context["cputune"][tag] = param
```
I wold move the `cputune` context into `context['cpu']['tuning']`

cbosdo commented on tests/unit/modules/test_virt.py:
```diff
@@ -2416,6 +2416,179 @@ def test_update(self):
+            setxml.find("./cputune/cachetune[@vcpus='0-3']/monitor[@vcpus='1']"), None
+        )
+        self.assertNotEqual(
+            setxml.find("./cputune/cachetune[@vcpus='4-5']").attrib.get("vcpus"), None
+        )
+        self.assertEqual(
+            setxml.find("./cputune/cachetune[@vcpus='4-5']/cache[@id='0']"), None
+        )
+        self.assertEqual(
+            setxml.find(
+                "./cputune/cachetune[@vcpus='4-5']/monitor[@vcpus='4']"
+            ).attrib.get("level"),
+            "3",
+        )
+        self.assertEqual(
+            setxml.find(
+                "./cputune/cachetune[@vcpus='4-5']/monitor[@vcpus='5']"
+            ).attrib.get("level"),
+            "2",
+        )
+        self.assertNotEqual(setxml.find("./cputune/memorytune[@vcpus='0-2']"), None)
+        self.assertEqual(
+            setxml.find("./cputune/memorytune[@vcpus='0-2']/node[@id='0']").attrib.get(
+                "bandwidth"
+            ),
+            "60",
+        )
+        self.assertNotEqual(setxml.find("./cputune/memorytune[@vcpus='3-4']"), None)
+        self.assertEqual(
+            setxml.find("./cputune/memorytune[@vcpus='3-4']/node[@id='0']").attrib.get(
+                "bandwidth"
+            ),
+            "50",
+        )
+        self.assertEqual(
+            setxml.find("./cputune/memorytune[@vcpus='3-4']/node[@id='1']").attrib.get(
+                "bandwidth"
+            ),
+            "70",
+        )
```
Would be nice to have such a test for `_gen_xml()` too, otherwise we don't test the XML generation, only the update

cbosdo reviewed (commented):
Another round of questions and things to enhance

cbosdo committed:
libvirt domain template memory config fixes

Add unit tests for _gen_xml() on the recently added memory parameters.
Also fixes an issue with an optional attribute.

cbosdo committed:
xmlutil.change_xml properly handle xpath node number

In XPath the node numbers are counted from 1 rather than 0.
Thus /foo/bar[0] is invalid and should be /foo/bar[1].

Since in the change_xml function we are getting the index from python
lists in these cases, we need to offset these.

cbosdo commented on salt/states/virt.py:
```diff
@@ -625,6 +776,30 @@ def running(
+        for instance: 'numatune': ``None``. Please note that ``None`` is mapped to ``null`` in sls file, pass ``null`` in
+        sls file instead.
+
+        .. versionadded:: Magnesium
```
```suggestion
        .. versionadded:: Aluminium
```

cbosdo reviewed (changes requested):
Still a `Magnesium` laying around, but the rest is OK

cbosdo reviewed (approved):
:+1:

cbosdo commented on salt/templates/virt/libvirt_domain.jinja:
```diff
@@ -1,32 +1,219 @@
+        <numatune>
+            {%- if 'memory' in numatune and numatune.memory %}
+                <memory mode='{{ numatune.memory.mode }}'
+                    {%- if numatune.memory.nodeset %}nodeset='{{ numatune.memory.nodeset }}'{%- endif %}
```
```suggestion
                    {%- if numatune.memory.nodeset %} nodeset='{{ numatune.memory.nodeset }}'{%- endif %}
```

There is a missing white space before the `nodeset` attribute that makes the XML invalid

cbosdo reviewed (changes requested):
Found a white space issue that breaks the XML validity

gqlo commented:
re-run centos

cbosdo reviewed (approved):
LGTM now

cbosdo commented:
@dwoz could you please merge this PR?

gqlo committed:
Merge branch 'master' into cpu_tune

gqlo committed:
Merge branch 'master' into cpu_tune

gqlo committed:
Merge branch 'master' into cpu_tune

gqlo committed:
Merge branch 'master' into cpu_tune

gqlo committed:
Merge branch 'master' into cpu_tune

gqlo committed:
Merge branch 'master' into cpu_tune

gqlo committed:
Merge branch 'master' into cpu_tune

gqlo committed:
Merge branch 'master' into cpu_tune

gqlo committed:
Merge branch 'master' into cpu_tune
